### PR TITLE
Query plan for creating sets

### PR DESCRIPTION
**Pull request by KochetovNicolai**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Show subqueries for `SET` and `JOIN` in `EXPLAIN` result.

Subqueries for set and join now use `QueryPlan` step as source. It also may be helpful for query optimisations.

### CODE REVIEW

**Comment by KochetovNicolai** [comment]
```sql
EXPLAIN
SELECT number
FROM numbers(10) AS f
ANY LEFT JOIN 
(
    SELECT number
    FROM numbers(10)
) AS s USING (number)
WHERE number IN 
(
    SELECT number * 2
    FROM numbers(10)
)
FORMAT TSV

Expression (Projection)
  CreatingSets (Create sets before main query execution)
    Expression (Before ORDER BY and SELECT)
      Filter (WHERE)
        Join (JOIN)
          Expression (Before JOIN)
            ReadFromStorage (Read from SystemNumbers)
    CreatingSet (Create set for subquery)
      Expression (Projection)
        Expression (Before ORDER BY and SELECT)
          ReadFromStorage (Read from SystemNumbers)
    CreatingSet (Create set for JOIN)
      Expression (Projection)
        Expression (Before ORDER BY and SELECT)
          ReadFromStorage (Read from SystemNumbers)
```

```sql
EXPLAIN PIPELINE
SELECT number
FROM numbers(10) AS f
ANY LEFT JOIN 
(
    SELECT number
    FROM numbers(10)
) AS s USING (number)
WHERE number IN 
(
    SELECT number * 2
    FROM numbers(10)
)
FORMAT TSV

(Expression)
ExpressionTransform × 2
  (CreatingSets)
  Concat 2 → 1
    (Expression)
    ExpressionTransform × 2
      (Filter)
      FilterTransform × 2
        (Join)
        JoiningTransform × 2
          SourceFromSingleChunk 0 → 1
            (Expression)
            ExpressionTransform
              (ReadFromStorage)
    (CreatingSet)
    CreatingSetsTransform
      (Expression)
      ExpressionTransform
        (Expression)
        ExpressionTransform
          (ReadFromStorage)
    (CreatingSet)
    CreatingSetsTransform
      (Expression)
      ExpressionTransform
        (Expression)
        ExpressionTransform
          (ReadFromStorage)

```

File: src/Processors/QueryPipeline.h [hunk 1]
```diff
@@ -87,6 +87,12 @@ class QueryPipeline
             size_t max_threads_limit = 0,
             Processors * collected_processors = nullptr);
 
+    /// Add other pipeline and execute it before current one.
+    /// Pipeline must have same header.
+    void addDelayingPipeline(QueryPipeline pipeline);
```
Comment (4ertus2) [inline src/Processors/QueryPipeline.h:92]: It's not clear which of them is delayed.  We could make a better name. Something like `addPipelineBefore()` or `addPipelineAfter()`.

File: src/Interpreters/SubqueryForSet.cpp [hunk 2]
```diff
@@ -8,13 +8,19 @@
 namespace DB
 {
 
+SubqueryForSet::SubqueryForSet() = default;
+SubqueryForSet::~SubqueryForSet() = default;
+SubqueryForSet::SubqueryForSet(SubqueryForSet &&) = default;
+SubqueryForSet & SubqueryForSet::operator= (SubqueryForSet &&) = default;
```
Comment (4ertus2) [inline src/Interpreters/SubqueryForSet.cpp:14]: Why do you place them here? What would go wrong if they would be defined in class?

File: src/Interpreters/GlobalSubqueriesVisitor.h [hunk 3]
```diff
@@ -135,7 +135,8 @@ class GlobalSubqueriesMatcher
                 ast = database_and_table_name;
 
             external_tables[external_table_name] = external_storage_holder;
-            subqueries_for_sets[external_table_name].source = QueryPipeline::getPipe(interpreter->execute().pipeline);
+            subqueries_for_sets[external_table_name].source = std::make_unique<QueryPlan>();
+            interpreter->buildQueryPlan(*subqueries_for_sets[external_table_name].source);
```
Comment (4ertus2) [inline src/Interpreters/GlobalSubqueriesVisitor.h:139]: It's the second place where we create `QueryPlan` just to `buildQueryPlan()` it at the next line. Would it better to make a combined function `interpreter->createQueryPlan()`?

File: src/Interpreters/InterpreterSelectQuery.cpp [hunk 4]
```diff
@@ -1899,14 +1899,11 @@ void InterpreterSelectQuery::executeSubqueriesInSetsAndJoins(QueryPlan & query_p
 
     const Settings & settings = context->getSettingsRef();
 
-    auto creating_sets = std::make_unique<CreatingSetsStep>(
-            query_plan.getCurrentDataStream(),
-            std::move(subqueries_for_sets),
-            SizeLimits(settings.max_rows_to_transfer, settings.max_bytes_to_transfer, settings.transfer_overflow_mode),
-            *context);
+    if (subqueries_for_sets.empty())
+        return;
```
Comment (4ertus2) [inline src/Interpreters/InterpreterSelectQuery.cpp:1903]: Could we have a situation with `query_info.input_order_info == true`  and empty subqueries_for_sets? If not it's a bit confucing to check it here. Probably we could make it for entire `executeSubqueriesInSetsAndJoins()` in call places.

File: src/Interpreters/InterpreterSelectWithUnionQuery.cpp [hunk 5]
```diff
@@ -183,13 +183,14 @@ void InterpreterSelectWithUnionQuery::buildQueryPlan(QueryPlan & query_plan)
...
-    std::vector<QueryPlan> plans(num_plans);
+    std::vector<std::unique_ptr<QueryPlan>> plans(num_plans);
     DataStreams data_streams(num_plans);
 
     for (size_t i = 0; i < num_plans; ++i)
     {
-        nested_interpreters[i]->buildQueryPlan(plans[i]);
-        data_streams[i] = plans[i].getCurrentDataStream();
+        plans[i] = std::make_unique<QueryPlan>();
+        nested_interpreters[i]->buildQueryPlan(*plans[i]);
```
Comment (4ertus2) [inline src/Interpreters/InterpreterSelectWithUnionQuery.cpp:192]: One another place for `createQuildPlan()` :)

File: src/Interpreters/SubqueryForSet.cpp (hunk 2 above)
Comment (KochetovNicolai) [reply src/Interpreters/SubqueryForSet.cpp:14]: Only `~SubqueryForSet()` must be implemented here cause `QueryPlan` is incomplete in header, and `~std::unique_ptr<QueryPlan>`  not always can be implemented. Everything else is here is just for consistency.

File: src/Interpreters/InterpreterSelectQuery.cpp (hunk 4 above)
Comment (KochetovNicolai) [reply src/Interpreters/InterpreterSelectQuery.cpp:1903]: Well, actually this is checked in both call places too, so `subqueries_for_sets` is never empty here. I will remove this check.

And for me it is actually strange that we sort something here

File: src/Interpreters/GlobalSubqueriesVisitor.h (hunk 3 above)
Comment (KochetovNicolai) [reply src/Interpreters/GlobalSubqueriesVisitor.h:139]: I will probably change an interface of `buildQueryPlan` later.

**Review by 4ertus2 — APPROVED** [review]

**Comment by KochetovNicolai** [comment]
`test_rename_distributed_parallel_insert_and_select` is not connected to this pr.

**Pull request merged**